### PR TITLE
Feature/declarative props

### PR DESCRIPTION
**Pull request by jodiedunlop**
# Declarative Props Feature

## Overview 

Declarative props provides a few features:

1. `withProps()` helper method
2. Ability to declare props on the transformer, without providing a `transform()` method.
3. Built-in formatters
4. Custom formatters

## withProps() helper

Within your transform method, simply call the withProps method with your model/data and a list of properties to expose:

```php
class EventTransformer extends AbstractTransformer
{
    public function transform(Event $event)
    {
        return $this->withProps($event, [
            'id',
            'receiver.id',
            'type',
            'content_type',
            'headers',
            'body',
            'is_verified',
            'created_at' => 'datetime',
            'updated_at' => 'datetime',
            'expires_at' => 'datetime',
        ]);
    }
}
```

- The properties are fetched from the given model (anything that implements ArrayAccess)
- Dot-notation properties like `receiver.id` are converted to a snake_case version eg. `receiver_id` to fetch the property from the model, and it is exposed as a nested array `"receiver" => [ "id" => "val" ]`

## Throw away your transform() method

In fact, with declarative props, you can simply declare a `$props` property:
```php
class EventTransformer extends AbstractTransformer
{
     protected $props = [
        'id',
        'receiver.id',
        'type',
        'content_type',
        'headers',
        'body',
        'is_verified',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
        'expires_at' => 'datetime',
    ];
}
```

When a transformer doesn't contain a `transform()` method, the resolver simply calls and returns `withProps($model, $this->getProps())` for you.

## Built-in formatters

A few default formatter types are included for casting/converting property values:

- int (int|integer)
- float (float|real|double)
- string
- bool (bool|boolean)
- array
- datetime
- date

You can apply these formatters by specifying a `type:formatter` within your property definition.

Eg. 

```php
 protected $props = [
        'id',
        'is_verified' => 'type:boolean',
        'created_at' => 'type:datetime',
        'expiry_date' => 'type:date',
    ];
```

All of the built-in types will allow you to omit the 'type:' prefix, so you can simply put:

```php
 protected $props = [
        'id',
        'is_verified' => 'boolean',
        'created_at' => 'datetime',
        'expiry_date' => 'date',
    ];
```

The `datetime` and `date` built-in types also support arguments for customising the behavior:

- `timezone:` - Convert the date relative to the timezone first
- `format:` - Provide an alternate format string

```php
 protected $props = [
        'created_at' => 'datetime|timezone:Australia/Brisbane',
        'expiry_date' => 'date|format:d/m/Y',
    ];
```

Note that you can also define the default formats and timezone by overriding the following properties in your Transformer class:

- `$dateFormat = 'Y-m-d'`
- `$dateTimeFormat = 'Y-m-d\TH:i:s.vP'`
- `$defaultTimezone = null`

## Custom formatters

Custom formatters can implemented by specifying a custom type in the definition, and a matching method:

```php
// Your transformer
 protected $props = [
        'dollaroos' => 'type:money',
    ];

protected function formatMoney($value, array $settings): string
{
    // We will also allow an optional format:... setting within the definition
    return money_format($settings['format'] ?? '%.2n', $value);
    
}
```

### Callback method

If you really want full control over how the property is resolved, you can provide an anonymous function for the definition:

```php
// Your transformer
public function getProps(): array
{
    return [
        'dollaroos' => function($model, $prop) {
            // $this is bound to the transformer
            return money_format('%.2n', $model->$prop);
        },
        'address' => function($model) {
            return sprintf("%s\n%s %s %s", $model->street, $model->suburb, $model->state, $model->postcode);
            }
    ];
}
```

- Your anonymous function is bound to the transformer, so you can still access `$this`

### CODE REVIEW

**Comment by ababkov** [comment]
Love it - especially the custom formatters ;)

Only one note - the parseDefinition method, maybe too generic a name when introduced in the context of a formatting trait?

**Comment by ababkov** [comment]
Oh yeah - can we add docs somewhere for posterity (pretty much everything you have in the notes for this PR - maybe in a docs folder and link to the topic from the main readme?)

**Comment by jodiedunlop** [comment]
Good point regarding parseDefinition naming. For now I will rename -  I plan to move this out into a simple class that can be used generically for both the transformer $includes definitions and the new declarative properties.

Re: docs, I've started working on smokescreen-docs, and mostly wrote this PR overview  knowing I'm going to move it into that docs site, since a README really isn't adequate for the scope of smokescreen functionality and practices.

**Comment by ababkov** [comment]
Cool on both notes.

File: tests/Unit/SmokescreenTest.php [hunk 1]
```diff
@@ -617,4 +732,18 @@ public function collection($resourceKey, array $data): array
             }
         };
     }
+
+    protected function createTransformerResolver(): TransformerResolverInterface
+    {
+        return new class() implements TransformerResolverInterface {
```
Comment (lachlankrautz) [inline tests/Unit/SmokescreenTest.php]: So java

**Pull request merged**